staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go

```diff
@@ -0,0 +1,264 @@
+	fairQueuingFactory FairQueuingFactory
+
+	// configQueue holds TypedConfigObjectReference values, identifying
+	// config objects that need to be processed
+	configQueue workqueue.RateLimitingInterface
+
+	// numConfigWorkers is the configured number of workers for the config queue
+	numConfigWorkers int
+
+	// plInformer is the informer for priority level config objects
+	plInformer cache.SharedIndexInformer
+
+	// fsInformer is the informer for flow schema config objects
+	fsInformer cache.SharedIndexInformer
+
+	// plLister belongs here too
+	// fsLister belongs here too
+
+	// serverConcurrencyLimit is the limit on the server's total
+	// number of non-exempt requests being served at once.  This comes
+	// from server configuration.
+	serverConcurrencyLimit int
+
+	// requestWaitLimit comes from server configuration.
+	requestWaitLimit time.Duration
+
+	// curState holds a pointer to the current RMState.  That is,
+	// `Load()` produces a `*RMState`.  When a config work queue worker
+	// processes a configuration change, it stores a new pointer here ---
+	// it does NOT side-effect the old `RMState` value.  The new `RMState`
+	// has a freshly constructed slice of FlowSchema pointers and a
+	// freshly constructed map of priority level states.  But the new
+	// `RMState.priorityLevelStates` includes in its range at least all the
+	// `*PriorityLevelState` values of the old `RMState.priorityLevelStates`.
+	// Consequently the filter can load a `*RMState` and work with it
+	// without concern for concurrent updates.  When a priority level is
+	// finally deleted, this will also involve storing a new `*RMState`
+	// pointer here, but in this case the range of the
+	// `RMState.priorityLevels` will be reduced --- by removal of the
+	// priority level that is no longer in use.
+	curState atomic.Value
+}
+
+// TypedConfigObjectReference is a reference to a relevant config API object.
+// No namespace is needed because none of these objects is namespaced.
+type TypedConfigObjectReference struct {
+	Kind string
+	Name string
+}
+
+func (tr *TypedConfigObjectReference) String() string {
+	return tr.Kind + "/" + tr.Name
+}
+
+// rmSetup is invoked at startup to create the infrastructure of this filter
+func rmSetup(loopbackClientConfig *restclient.Config, serverConcurrencyLimit int, requestWaitLimit time.Duration) *requestManagement {
+	reqMgmt := &requestManagement{
+		configQueue:            workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 8*time.Hour), "req_mgmt_config_queue"),
+		serverConcurrencyLimit: serverConcurrencyLimit,
+		requestWaitLimit:       requestWaitLimit,
+	}
+	// TODO: finish implementation
+	return reqMgmt
+}
+
+// WithRequestManagement limits the number of in-flight requests in a fine-grained way
+func WithRequestManagement(
+	handler http.Handler,
+	loopbackClientConfig *restclient.Config,
+	serverConcurrencyLimit int,
+	requestWaitLimit time.Duration,
```

Comment:
`requestWaitLimit` is this meant for a time-out for those requests pending in queues? there's already a time-out handler filter, we probably don't want them to overlap..

Reply:
This parameter is introduced in https://github.com/kubernetes/enhancements/blob/79119ce549cda9cdfdb532273475ea5090fa5415/keps/sig-api-machinery/20190228-priority-and-fairness.md#latency-protection .  In particular, I think it must be set to something substantially shorter than the limit imposed by the timeout filter.  It does not make sense to allow a request to wait in a queue for longer than the total time allowed for waiting+executing.  In fact, I think the wait duration limit should be short enough that the majority of the overall duration limit is available for request execution.

Reply:
>  I think it must be set to something substantially shorter than the limit imposed by the timeout filter

i see, we can check that when validating the flags, what do you think will be the default value for this timeout?

Reply:
I suggest scaling it with the upstream timeout.  I suggest setting the wait duration limit to 1/4 of the upstream timeout.